src/components/Canvas.tsx

```diff
@@ -52,15 +55,29 @@ const Canvas: React.VFC<Props> = (props) => {
 
   const handleMouseUp = () => {
     isDrawing.current = false;
+    setLines(lines.slice(0, historyStep + 1));
+    setHistory(history.slice(0, historyStep).concat([lines.slice()]));
+    setHistoryStep(historyStep + 1);
   };
 
   const handleChangeToolType = (type: ToolType) => {
     setTool(type);
   };
 
+  const handleUndo = () => {
+    if (historyStep === 0) {
+      return;
+    }
+    setHistoryStep(historyStep - 1);
+    setLines(lines.slice(0, historyStep - 1));
```

Comment:
提案です。以下の記述はどうでしょうか？

```suggestion
    setLines(history[historyStep - 1]);
```

tenさんのコードとの違いは以下になります。
- tenさん：現在のキャンバスから最新の一手を取り消した、新たな配列を作成する
- 提案：historyStepを使用し、historyの一手前を参照している

履歴をさかのぼるといった文脈にあった表現を考えると、こうしたほうがわかりやすいかもしれません。